### PR TITLE
[QUIC] API QuicStream

### LINKED ISSUE #69675: [API Proposal]: [QUIC] QuicStream
### Background and motivation

API design for exposing `QuicStream` and related classes to the public.

The API shape is mainly defined by `Stream` from which it derives. `QuicStream` has some additional APIs specific for QUIC, but it still has a goal of `QuicStream` to be usable as an ordinary stream, e.g.: allow usage like this:
```csharp
await using var stream = ...; // Get QuicStream
await stream.WriteAsync(...);
await stream.ReadAsync(...);
// Dispose by await using
```
Albeit better results can be achieved via the additional API. For example from the previous code snippet, the stream didn't complete writes until `DisposeAsync` got called. Thus the peer might still be expecting incoming data and never send any itself. All of this depends on the specific protocol and the contract between the peers. For instance in HTTP/3 case, we wrap `QuicStream` into HTTP/3 specific stream taking care of all the proprieties with completing the write side of the stream, which is then handed out in the response.

This API proposal concentrates on the additional, QUIC specific APIs.

#### Related issues:
- https://github.com/dotnet/runtime/issues/57780
- https://github.com/dotnet/runtime/issues/756
- https://github.com/dotnet/runtime/issues/43290
- https://github.com/dotnet/runtime/issues/55816

### API Proposal

```csharp
namespace System.Net.Quic;

public sealed class QuicStream : Stream
{
    // Stream API implied, QUIC specifics follow:

    public long Id { get; } // https://www.rfc-editor.org/rfc/rfc9000.html#name-stream-types-and-identifier
    public QuicStreamType Type { get; }  // https://github.com/dotnet/runtime/issues/55816, not necessary per se, CanRead and CanWrite should suffice

    // In 6.0 we had bool ReadsCompleted (tailored for ASP.NET) that would get set only in graceful case. The task gives the ability to distinguish error cases and is analogous to WritesCompleted.
    public Task ReadsClosed { get; } // gets set when STREAM frame with FIN bit (=EOF, =ReadAsync returning 0) is received or when the peer aborts the sending side by sending RESET_STREAM frame. Inspired by channel - might be ValueTask.
    
    // In 6.0 we had a method Task WaitForWriteCompletionAsync(). We need a Task that is removed from the operation kicking the write completion but gets completed when the sending side gets closed (either by CompleteWrites, endStream=true, Abort(Write) or Abort(Read) from the peer).
    public Task WritesClosed { get; } // gets set when the peer acknowledges STREAM frame with FIN bit (=EOF) or RESET_STREAM frame or when the peer aborts the receiving side by sending STOP_SENDING frame. Inspired by channel - might be ValueTask.

    // In 6.0 we had separate methods AbortRead and AbortWrite. This allows aborting both directions at the same time. It can remain split, it's not strictly necessary to have it combined.
    public void Abort(QuicAbortDirection abortDirection, long errorCode); // abortively ends either sending or receiving or both sides of the stream, i.e.: RESET_STREAM frame or STOP_SENDING frame

    // In 6.0 we had void Shutdown(), it was really badly named. The new name comes from DuplexStream API review, where CompleteWrites was chosen,
    public void CompleteWrites(); // https://github.com/dotnet/runtime/issues/43290, gracefully ends the sending side, equivalent to WriteAsync with endStream set to true.

    // Overload with endStream.
    public ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, bool completeWrites, CancellationToken cancellationToken = default);
}

[Flags]
enum QuicAbortDirection
{
   Read = 1,
   Write = 2,
   Both = Read | Write
}

public enum QuicStreamType
{
    Unidirectional,
    Bidirectional
}
```

### Client usage:
```C#
// Consider connection from https://github.com/dotnet/runtime/issues/68902:
await using var stream = await connection.OpenStreamAsync(QuicStreamType.Bidirectional, cancellationToken);
// Send some data.
await stream.WriteAsync(data, cancellationToken);
await stream.WriteAsync(data, cancellationToken);

// End the sending-side.
await stream.WriteAsync(data, endStream: true, cancellationToken);
// or
await stream.WriteAsync(data, cancellationToken);
stream.CompleteWrites();

// Read data until the end of stream.
while (await stream.ReadAsync(buffer, cancellationToken) > 0)
{
    // Handle buffer data...
}

// DisposeAsync called by await using.
```

### Server usage:
```C#
// Consider connection from https://github.com/dotnet/runtime/issues/68902:
await using var stream = await connection.AcceptStreamAsync(cancellationToken);

// Read the data.
while (stream.ReadAsync(buffer, cancellationToken) > 0)
{
    // Handle buffer data...

    // Peer send FIN flag with the last read.
    if (stream.ReadsCompleted.IsCompleted)
    {
        break;
    }
}

// Listen for Abort(Read) (STOP_SENDING) from the peer.
var writesCompletedTask = WritesCompletedAsync(stream);
async ValueTask WritesCompletedAsync(QuicStream stream)
{
    try
    {
        await stream.WritesCompleted;
    }
    catch (Exception ex)
    {
        // Abort the stream, peer send STOP_SENDING
    }
}

// DisposeAsync called by await using.
```

#### Read is finished
https://github.com/dotnet/runtime/issues/57780
Currently we have a `bool` property, tailored to ASP.NET core needs, the property gets set only on graceful completion of the peer's sending side.
```csharp
// Original. If it works... :)
public bool ReadsCompleted { get; }

// Proposal covers also abortive completions with details
public Task ReadsCompleted { get; } 
```
#### Sending EOF, getting notified about EOF
https://github.com/dotnet/runtime/issues/43290
Currently we have:
```csharp
// Gracefully ends the sending side, equivalent to endStream argument of WriteAsync
public void Shutdown();
// Doesn't kick off any operation leading to stream shutdown (for that Shutdown or Abort must be called). 
// Returns Task that gets completed when we gracefully end sending or we receive STOP_SENDING (i.e. peer called AbortRead).
public ValueTask WaitForWriteCompletionAsync(CancellationToken cancellationToken);
```

In #43290 (https://github.com/dotnet/runtime/issues/43290#issuecomment-820633348), this relevant part of API was approved:
```csharp
// Shutdown renamed, in this proposal we have the same API.
public abstract void CompleteWrites();
// Task WritesCompleted property equivalent. Looks like async version of CompleteWrites but isn't, since it doesn't kick off any operation leading to stream sending completion.
public abstract ValueTask CompleteWritesAsync(CancellationToken cancellationToken = default);
```

#### Abortive completion
https://github.com/dotnet/runtime/issues/756
Currently we have 2 methods:
```csharp
public void AbortRead(long errorCode);
public void AbortWrite(long errorCode);
```

In #756 (https://github.com/dotnet/runtime/issues/756#issuecomment-823442692) we agreed on what this proposal has:
```csharp
public void Abort(QuicAbortDirection abortDirection, long errorCode);
```

#### Stream type
https://github.com/dotnet/runtime/issues/55816
We can use `CanRead` and `CanWrite` from stream as we do now, no need for any additional API. The only disadvantage is that they both return `false` after disposal, thus are not useful at that time. Whether that's a valid / expected usage is a different question.

### Risks

As I'll state with all QUIC APIs. We might consider making all of these PreviewFeature. Not to deter user from using it, but to give us flexibility to tune the API shape based on customer feedback.
We don't have many users now and we're mostly making these APIs based on what Kestrel needs, our limited experience with System.Net.Quic and my meager experiments with other QUIC implementations.

### PR DESCRIPTION
**Pull request by ManickaP**
Fixes https://github.com/dotnet/runtime/issues/69675

~This PR is based on https://github.com/dotnet/runtime/pull/71783 so it shows more changes than it actually contains, see https://github.com/ManickaP/runtime/compare/mapichov/quic-connection...ManickaP:runtime:mapichov/quic-stream?expand=1 for the actual diff.~

ATM depends on: https://github.com/microsoft/msquic/pull/2883 but AFAIK only one outerloop test is failing because of this (the one testing stream shutdown due to idle connection). Depending on severity of this and speed on that msquic PR, I may temporarily re-introduce connection state from:
https://github.com/dotnet/runtime/blob/b48084582ea55b137fa043d3a3b7fac5b6b85285/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs#L120-L125

Also I'll be filling in comments, not everything is covered.

Kept in draft until API is approved.

### CODE REVIEW

File: src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs
Comment (CarnaViire) [inline src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs:988]: Should that be `IsCompletedSuccessfully`? in all places where we previously checked `ReadsCompleted` for `true`

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs
Comment (CarnaViire) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:219]: Why is this inside do...while ? One iteration can copy, and the following can throw?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs
Comment (CarnaViire) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:577]: How do we make sure send buffers are not disposed in the middle of send operation?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:219]: The loop can iterate at most twice. First, it tries to copy from the buffer, if it succeeds (data are already there) it will break out of the loop in `while (... totalCopied == 0);`. If no data are available in the buffer, the `await valueTask.ConfigureAwait(false);` will wait on the next `RECEIVE` event and once it arrives it will repeat the loop with copying, copy something (or learn about FIN flag) and break out of the loop.

The `valueTask` must be awaited each time it's retrieved, even if it's completed and it seems pointless - e.g.: `_receiveTcs.TrySetResult();`, since it's backed up by MRVTS which keeps "version" and the `await` will reset it (increment it).

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:577]: By awaiting the `valueTask` above which waits on `SHUTDOWN_COMPLETE` which is the last event on the stream and means both side are closed, so no send operation is in progress.

File: src/libraries/Common/tests/System/Net/Http/Http2LoopbackServer.cs
Comment (wfurt) [inline src/libraries/Common/tests/System/Net/Http/Http2LoopbackServer.cs:151]: I don't understand what this is trying to achieve.

File: src/libraries/Common/tests/System/Net/Http/Http3LoopbackServer.cs
Comment (wfurt) [inline src/libraries/Common/tests/System/Net/Http/Http3LoopbackServer.cs:92]: nit: `Http3LoopbackConnection con` instead of `var`

File: src/libraries/Common/tests/System/Net/Http/Http3LoopbackStream.cs
Comment (wfurt) [inline src/libraries/Common/tests/System/Net/Http/Http3LoopbackStream.cs:285]: We no longer do the `ShutdownCompleted`  because we expect that in Dispose? Is there reason to delay that if we know this is final?

**Review by wfurt — APPROVED** [review]
Generally looks ok to me. I wish we have better names for some of the handles but that is ok. 
It seems like we also break the build/tests in few places and that should be fixed before merging.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/msquic_generated.cs
Comment (rzikm) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/msquic_generated.cs:2414]: I don't see this change in the original MsQuic sources nor the C equivalent.

File: src/libraries/Common/tests/System/Net/Http/Http3LoopbackServer.cs
Comment (ManickaP) [reply src/libraries/Common/tests/System/Net/Http/Http3LoopbackServer.cs:92]: It was like this before, I just added the `await`. If we want to address `var` vs. concrete type, we can do that as a separate clean up PR afterwards, wdyt?

File: src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.Authentication.cs
Comment (ManickaP) [reply src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.Authentication.cs:720]: It was like this before, I just added the `await`. If we want to address `var` vs. concrete type, we can do that as a separate clean up PR afterwards, wdyt?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/msquic_generated.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/msquic_generated.cs:2414]: I'll remove it and return the `internal sealed class State` for now.

BTW, it's in this PR https://github.com/microsoft/msquic/pull/2883. I haven't had the chance to finish it (adding tests on Nick's request), but I do plan to publish it and propagate to dotnet/msquic and our docker images within 7.0 time frame.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs
Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:219]: A comment like "this will loop at most twice" would be nice :)

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs
Comment (rzikm) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:344]: I added string resources for theses.
```suggestion
            if (_receiveTcs.TrySetException(ThrowHelper.GetOperationAbortedException(SR.net_quic_reading_aborted), final: true))
```

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs
Comment (rzikm) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:351]:
```suggestion
            if (_sendTcs.TrySetException(ThrowHelper.GetOperationAbortedException(SR.net_quic_writing_aborted), final: true))
```

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs
Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:577]: Can you add a comment for this please? This is something I am going to forget very soon :D

File: src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs
Comment (rzikm) [inline src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs:916]:
```suggestion
                QuicException ex = AssertThrowsQuicExceptionAsync(QuicError.ConnectionAborted, () => clientStream.ReadAsync(buffer).AsTask());
                Assert.Equal(ExpectedErrorCode, ex.ApplicationErrorCode);
```

**Review by rzikm — APPROVED** [review]
Looks good modulo existing comments

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs:457]: Maybe we don't need this anymore. The original issue was that we're throwing `QuicOperationAborted` for shutdown by transport, but Kestrel expected that exception only for local connection close.
Now, we'd throw `QuicException` with `ConnectionIdle` for instance, so it's probably fine. Yeah, I think we can use the same exception, good catch (y)

File: src/libraries/Common/tests/System/Net/Http/Http3LoopbackServer.cs
Comment (ManickaP) [reply src/libraries/Common/tests/System/Net/Http/Http3LoopbackServer.cs:92]: Changing this particular case, but I'm not digging for all the other occurrences in this PR.

File: src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.Authentication.cs
Comment (ManickaP) [reply src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.Authentication.cs:720]: Changing this particular case, but I'm not digging for all the other occurrences in this PR.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:351]: Bad merge, fixing.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:344]: Bad merge, fixing.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:222]: Done.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs]: Changed.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/msquic_generated.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/msquic_generated.cs]: Done.

**Comment by ManickaP** [comment]
All comments should be answered, only waiting on CI.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs
Comment (CarnaViire) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:229]: I still don't understand how we would overcome the race when - between the first and the second iteration - a parallel read could come and "steal" the arrived data? What would happen to the first read? It seems like it could either throw or - would it start waiting on data to arrive again?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs
Comment (CarnaViire) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:253]: Do I understand correctly, that `complete` can *only* become true in "synchronous" case? Meaning, we didn't actually wait for anything in valueTask above

**Comment by ManickaP** [comment]
One failure unrelated: https://github.com/dotnet/runtime/issues/71233
One failure related (windows), reproduced locally, fixed and tested locally in a tight loop.
--> Merging now.

**Comment by ManickaP** [comment]
@JamesNK this will need your reaction as well. This contains the stream changes, most interesting will be the ReadsClosed, WritesClosed tasks.

**Pull request merged**
